File: stacks/python/base/code/http-client.tpl.py
# Universal Template System - Python Stack
# Generated: 2025-12-10
# Purpose: HTTP client utilities
# Tier: base
# Stack: python
# Category: utilities

#!/usr/bin/env python3
# -----------------------------------------------------------------------------
# FILE: http-client.tpl.py
# PURPOSE: Comprehensive HTTP client utilities for Python projects
# USAGE: Import and adapt for consistent HTTP communication across the application
# DEPENDENCIES: requests, json, logging for HTTP operations and response handling
# AUTHOR: [[.Author]]
# VERSION: [[.Version]]
# SINCE: [[.Version]]
# -----------------------------------------------------------------------------

"""
Python HTTP Client Utilities Template
Purpose: Reusable HTTP client utilities for Python projects
Usage: Import and adapt for consistent HTTP communication across the application
"""

import requests
import json
import logging
from typing import Dict, Any, Optional, Union, List
from dataclasses import dataclass
from urllib.parse import urljoin
import time
from enum import Enum

class HTTPMethod(Enum):
    """HTTP methods enumeration"""
    GET = "GET"
    POST = "POST"
    PUT = "PUT"
    DELETE = "DELETE"
    PATCH = "PATCH"
    HEAD = "HEAD"
    OPTIONS = "OPTIONS"

@dataclass
class HTTPResponse:
    """HTTP response wrapper"""
    status_code: int
    data: Any
    headers: Dict[str, str]
    success: bool
    error: Optional[str] = None
    response_time: Optional[float] = None

@dataclass
class HTTPRequest:
    """HTTP request configuration"""
    method: HTTPMethod
    url: str
    headers: Optional[Dict[str, str]] = None
    params: Optional[Dict[str, Any]] = None
    data: Optional[Union[Dict, str, bytes]] = None
    json_data: Optional[Dict[str, Any]] = None
    timeout: Optional[float] = 30.0
    retries: int = 3

class HTTPClientError(Exception):
    """HTTP client error"""
    
    def __init__(self, message: str, status_code: int = None, response_data: Any = None):
        super().__init__(message)
        self.status_code = status_code
        self.response_data = response_data

class HTTPClient:
    """Reusable HTTP client with retry logic and error handling"""
    
    def __init__(
        self,
        base_url: str = None,
        default_headers: Optional[Dict[str, str]] = None,
        timeout: float = 30.0,
        retries: int = 3,
        logger: Optional[logging.Logger] = None
    ):
        self.base_url = base_url
        self.default_headers = default_headers or {}
        self.timeout = timeout
        self.retries = retries
        self.logger = logger or logging.getLogger(__name__)
        
        # Create session for connection pooling
        self.session = requests.Session()
        self.session.headers.update(self.default_headers)
    
    def _build_url(self, endpoint: str) -> str:
        """Build full URL from base URL and endpoint"""
        if self.base_url:
            return urljoin(self.base_url, endpoint)
        return endpoint
    
    def _make_request(self, request: HTTPRequest) -> HTTPResponse:
        """Make HTTP request with retry logic"""
        
        start_time = time.time()
        last_exception = None
        
        for attempt in range(request.retries + 1):
            try:
                # Prepare request arguments
                kwargs = {
                    'timeout': request.timeout,
                    'headers': request.headers or {}
                }
                
                if request.params:
                    kwargs['params'] = request.params
                
                if request.json_data:
                    kwargs['json'] = request.json_data
                elif request.data:
                    kwargs['data'] = request.data
                
                # Make request
                response = self.session.request(
                    method=request.method.value,
                    url=self._build_url(request.url),
                    **kwargs
                )
                
                # Calculate response time
                response_time = time.time() - start_time
                
                # Parse response
                try:
                    if response.headers.get('content-type', '').startswith('application/json'):
                        data = response.json()
                    else:
                        data = response.text
                except (json.JSONDecodeError, ValueError):
                    data = response.text
                
                # Create response object
                http_response = HTTPResponse(
                    status_code=response.status_code,
                    data=data,
                    headers=dict(response.headers),
                    success=200 <= response.status_code < 300,
                    response_time=response_time
                )
                
                # Log request
                self._log_request(request, http_response, attempt + 1)
                
                # Check for HTTP errors
                if not http_response.success:
                    raise HTTPClientError(
                        f"HTTP {response.status_code}: {response.reason}",
                        status_code=response.status_code,
                        response_data=data
                    )
                
                return http_response
                
            except requests.exceptions.RequestException as e:
                last_exception = e
                if attempt < request.retries:
                    self.logger.warning(f"Request failed (attempt {attempt + 1}), retrying: {str(e)}")
                    time.sleep(2 ** attempt)  # Exponential backoff
                else:
                    self.logger.error(f"Request failed after {attempt + 1} attempts: {str(e)}")
        
        # All retries failed
        raise HTTPClientError(f"Request failed after {request.retries + 1} attempts: {str(last_exception)}")
    
    def _log_request(self, request: HTTPRequest, response: HTTPResponse, attempt: int):
        """Log HTTP request and response"""
        self.logger.info(
            f"HTTP {request.method.value} {request.url} -> {response.status_code} "
            f"({response.response_time:.3f}s, attempt {attempt})"
        )
    
    def get(self, url: str, params: Optional[Dict] = None, **kwargs) -> HTTPResponse:
        """Make GET request"""
        request = HTTPRequest(method=HTTPMethod.GET, url=url, params=params, **kwargs)
        return self._make_request(request)
    
    def post(self, url: str, data: Optional[Dict] = None, json_data: Optional[Dict] = None, **kwargs) -> HTTPResponse:
        """Make POST request"""
        request = HTTPRequest(method=HTTPMethod.POST, url=url, data=data, json_data=json_data, **kwargs)
        return self._make_request(request)
    
    def put(self, url: str, data: Optional[Dict] = None, json_data: Optional[Dict] = None, **kwargs) -> HTTPResponse:
        """Make PUT request"""
        request = HTTPRequest(method=HTTPMethod.PUT, url=url, data=data, json_data=json_data, **kwargs)
        return self._make_request(request)
    
    def delete(self, url: str, **kwargs) -> HTTPResponse:
        """Make DELETE request"""
        request = HTTPRequest(method=HTTPMethod.DELETE, url=url, **kwargs)
        return self._make_request(request)
    
    def patch(self, url: str, data: Optional[Dict] = None, json_data: Optional[Dict] = None, **kwargs) -> HTTPResponse:
        """Make PATCH request"""
        request = HTTPRequest(method=HTTPMethod.PATCH, url=url, data=data, json_data=json_data, **kwargs)
        return self._make_request(request)

class APIClient:
    """Base class for specific API clients"""
    
    def __init__(self, base_url: str, api_key: str = None, **kwargs):
        self.client = HTTPClient(base_url=base_url, **kwargs)
        self.api_key = api_key
        
        # Set authentication headers if API key provided
        if api_key:
            self.client.session.headers.update({
                'Authorization': f'Bearer {api_key}',
                'Content-Type': 'application/json'
            })
    
    def _handle_response(self, response: HTTPResponse) -> Any:
        """Handle API response with common error patterns"""
        if not response.success:
            if response.status_code == 401:
                raise HTTPClientError("Authentication failed", response.status_code)
            elif response.status_code == 403:
                raise HTTPClientError("Access denied", response.status_code)
            elif response.status_code == 404:
                raise HTTPClientError("Resource not found", response.status_code)
            elif response.status_code >= 500:
                raise HTTPClientError("Server error", response.status_code)
            else:
                raise HTTPClientError(f"API error: {response.data}", response.status_code)
        
        return response.data

# Utility functions for common HTTP patterns
def download_file(url: str, file_path: str, chunk_size: int = 8192) -> bool:
    """Download file from URL"""
    try:
        response = requests.get(url, stream=True)
        response.raise_for_status()
        
        with open(file_path, 'wb') as f:
            for chunk in response.iter_content(chunk_size=chunk_size):
                f.write(chunk)
        
        return True
    except Exception as e:
        logging.error(f"Failed to download file from {url}: {e}")
        return False

def upload_file(url: str, file_path: str, field_name: str = 'file', **kwargs) -> HTTPResponse:
    """Upload file to URL"""
    client = HTTPClient()
    
    with open(file_path, 'rb') as f:
        files = {field_name: f}
        request = HTTPRequest(method=HTTPMethod.POST, url=url, data=kwargs, files=files)
        return client._make_request(request)

def batch_requests(requests: List[HTTPRequest], client: HTTPClient = None) -> List[HTTPResponse]:
    """Execute multiple HTTP requests concurrently"""
    if client is None:
        client = HTTPClient()
    
    import concurrent.futures
    
    def execute_request(request):
        return client._make_request(request)
    
    with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
        futures = [executor.submit(execute_request, req) for req in requests]
        responses = [future.result() for future in concurrent.futures.as_completed(futures)]
    
    return responses

# Example usage
if __name__ == "__main__":
    import logging
    
    # Setup logging
    logging.basicConfig(level=logging.INFO)
    logger = logging.getLogger(__name__)
    
    # Create HTTP client
    client = HTTPClient(
        base_url="https://jsonplaceholder.typicode.com",
        default_headers={"User-Agent": "MyApp/1.0"},
        timeout=10.0,
        logger=logger
    )
    
    # Test GET request
    try:
        response = client.get("/posts/1")
        print(f"GET request successful: {response.data}")
    except HTTPClientError as e:
        print(f"GET request failed: {e}")
    
    # Test POST request
    try:
        new_post = {
            "title": "Test Post",
            "body": "This is a test post",
            "userId": 1
        }
        response = client.post("/posts", json_data=new_post)
        print(f"POST request successful: {response.data}")
    except HTTPClientError as e:
        print(f"POST request failed: {e}")
    
    # Test API client
    class JSONPlaceholderAPI(APIClient):
        def get_posts(self) -> List[Dict]:
            return self._handle_response(self.client.get("/posts"))
        
        def create_post(self, post_data: Dict) -> Dict:
            return self._handle_response(self.client.post("/posts", json_data=post_data))
    
    api = JSONPlaceholderAPI("https://jsonplaceholder.typicode.com")
    
    try:
        posts = api.get_posts()
        print(f"Retrieved {len(posts)} posts")
        
        new_post = api.create_post({
            "title": "API Test Post",
            "body": "Created via API client",
            "userId": 1
        })
        print(f"Created post: {new_post}")
        
    except HTTPClientError as e:
        print(f"API request failed: {e}")
    
    print("HTTP client utilities demo completed")

File: stacks/go/base/code/http-client.tpl.go
// Template: http-client.tpl.go
// Purpose: http-client template
// Stack: go
// Tier: base

# Universal Template System - Go Stack
# Generated: 2025-12-10
# Purpose: HTTP client utilities
# Tier: base
# Stack: go
# Category: utilities

// -----------------------------------------------------------------------------
// FILE: http-client.tpl.go
// PURPOSE: Comprehensive HTTP client utilities for Go projects
// USAGE: Import and adapt for consistent HTTP communication across the application
// DEPENDENCIES: bytes, context, encoding/json, fmt, io, net/http, net/url
// AUTHOR: [[.Author]]
// VERSION: [[.Version]]
// SINCE: [[.Version]]
// -----------------------------------------------------------------------------

package httpclient

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// HTTPMethod represents HTTP methods
type HTTPMethod string

const (
	GET    HTTPMethod = "GET"
	POST   HTTPMethod = "POST"
	PUT    HTTPMethod = "PUT"
	DELETE HTTPMethod = "DELETE"
	PATCH  HTTPMethod = "PATCH"
	HEAD   HTTPMethod = "HEAD"
)

// HTTPResponse represents an HTTP response
type HTTPResponse struct {
	StatusCode    int                    `json:"status_code"`
	Data          interface{}            `json:"data"`
	Headers       map[string]string      `json:"headers"`
	Success       bool                   `json:"success"`
	Error         string                 `json:"error,omitempty"`
	ResponseTime  time.Duration          `json:"response_time,omitempty"`
	RequestID     string                 `json:"request_id,omitempty"`
	ContentType   string                 `json:"content_type,omitempty"`
	RawBody       []byte                 `json:"-"`
}

// HTTPClientError represents an HTTP client error
type HTTPClientError struct {
	Message      string        `json:"message"`
	StatusCode   int           `json:"status_code"`
	ResponseData interface{}   `json:"response_data,omitempty"`
	RequestID    string        `json:"request_id,omitempty"`
	ResponseTime time.Duration `json:"response_time,omitempty"`
}

// Error implements the error interface
func (e *HTTPClientError) Error() string {
	return e.Message
}

// HTTPClientConfig represents HTTP client configuration
type HTTPClientConfig struct {
	BaseURL         string            `json:"base_url"`
	DefaultHeaders  map[string]string `json:"default_headers"`
	Timeout         time.Duration     `json:"timeout"`
	MaxRetries      int               `json:"max_retries"`
	RetryDelay      time.Duration     `json:"retry_delay"`
	EnableLogging   bool              `json:"enable_logging"`
	EnableCache     bool              `json:"enable_cache"`
	CacheMaxAge     time.Duration     `json:"cache_max_age"`
	UserAgent       string            `json:"user_agent"`
	FollowRedirects bool              `json:"follow_redirects"`
}

// DefaultHTTPClientConfig returns default HTTP client configuration
func DefaultHTTPClientConfig() HTTPClientConfig {
	return HTTPClientConfig{
		DefaultHeaders: map[string]string{
			"Content-Type": "application/json",
			"Accept":       "application/json",
		},
		Timeout:         30 * time.Second,
		MaxRetries:      3,
		RetryDelay:      1 * time.Second,
		EnableLogging:   false,
		EnableCache:     false,
		CacheMaxAge:     5 * time.Minute,
		UserAgent:       "Go-HTTPClient/1.0",
		FollowRedirects: true,
	}
}

// HTTPClient represents an HTTP client
type HTTPClient struct {
	config     HTTPClientConfig
	httpClient *http.Client
	cache      *RequestCache
	metrics    *HTTPMetrics
	interceptors []RequestInterceptor
}

// RequestInterceptor interface for request/response interceptors
type RequestInterceptor interface {
	Intercept(req *HTTPRequest) (*HTTPRequest, error)
}

// HTTPRequest represents an HTTP request
type HTTPRequest struct {
	Method      HTTPMethod         `json:"method"`
	URL         string             `json:"url"`
	Headers     map[string]string  `json:"headers"`
	Body        interface{}        `json:"body,omitempty"`
	QueryParams map[string]string  `json:"query_params,omitempty"`
	Context     context.Context    `json:"-"`
	RequestID   string             `json:"request_id,omitempty"`
	Timeout     time.Duration      `json:"timeout,omitempty"`
}

// NewHTTPClient creates a new HTTP client
func NewHTTPClient(config HTTPClientConfig) *HTTPClient {
	client := &HTTPClient{
		config: config,
		httpClient: &http.Client{
			Timeout: config.Timeout,
		},
		cache:      NewRequestCache(config.CacheMaxAge),
		metrics:    NewHTTPMetrics(),
		interceptors: make([]RequestInterceptor, 0),
	}

	// Set user agent
	if config.UserAgent != "" {
		client.config.DefaultHeaders["User-Agent"] = config.UserAgent
	}

	return client
}

// AddInterceptor adds a request interceptor
func (c *HTTPClient) AddInterceptor(interceptor RequestInterceptor) {
	c.interceptors = append(c.interceptors, interceptor)
}

// buildURL builds the full URL
func (c *HTTPClient) buildURL(path string) string {
	if strings.HasPrefix(path, "http://") || strings.HasPrefix(path, "https://") {
		return path
	}
	return c.config.BaseURL + path
}

// buildHeaders builds headers for the request
func (c *HTTPClient) buildHeaders(additionalHeaders map[string]string) map[string]string {
	headers := make(map[string]string)
	for k, v := range c.config.DefaultHeaders {
		headers[k] = v
	}
	for k, v := range additionalHeaders {
		headers[k] = v
	}
	return headers
}

// generateRequestID generates a unique request ID
func (c *HTTPClient) generateRequestID() string {
	return fmt.Sprintf("req-%d", time.Now().UnixNano())
}

// executeRequest executes an HTTP request with retry logic
func (c *HTTPClient) executeRequest(req *HTTPRequest) (*HTTPResponse, error) {
	var lastErr error
	
	for attempt := 0; attempt <= c.config.MaxRetries; attempt++ {
		response, err := c.doRequest(req)
		if err == nil {
			return response, nil
		}

		lastErr = err

		if attempt < c.config.MaxRetries && c.shouldRetry(err) {
			delay := c.config.RetryDelay * time.Duration(attempt+1)
			time.Sleep(delay)
			continue
		}

		break
	}

	return nil, lastErr
}

// doRequest performs a single HTTP request
func (c *HTTPClient) doRequest(req *HTTPRequest) (*HTTPResponse, error) {
	startTime := time.Now()

	// Apply interceptors
	for _, interceptor := range c.interceptors {
		modifiedReq, err := interceptor.Intercept(req)
		if err != nil {
			return nil, err
		}
		req = modifiedReq
	}

	// Check cache for GET requests
	if req.Method == GET && c.config.EnableCache {
		if cached := c.cache.Get(req.URL, req.Headers); cached != nil {
			return cached, nil
		}
	}

	// Build URL with query parameters
	fullURL := req.URL
	if len(req.QueryParams) > 0 {
		u, err := url.Parse(req.URL)
		if err != nil {
			return nil, fmt.Errorf("invalid URL: %w", err)
		}
		
		q := u.Query()
		for k, v := range req.QueryParams {
			q.Set(k, v)
		}
		u.RawQuery = q.Encode()
		fullURL = u.String()
	}

	// Prepare request body
	var bodyReader io.Reader
	if req.Body != nil {
		bodyBytes, err := json.Marshal(req.Body)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal request body: %w", err)
		}
		bodyReader = bytes.NewReader(bodyBytes)
	}

	// Create HTTP request
	httpReq, err := http.NewRequestWithContext(req.Context, string(req.Method), fullURL, bodyReader)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	// Set headers
	for k, v := range req.Headers {
		httpReq.Header.Set(k, v)
	}

	// Set request ID header
	if req.RequestID != "" {
		httpReq.Header.Set("X-Request-ID", req.RequestID)
	}

	// Execute request
	httpResp, err := c.httpClient.Do(httpReq)
	if err != nil {
		responseTime := time.Since(startTime)
		c.metrics.RecordError(req.Method, err)
		return nil, &HTTPClientError{
			Message:      err.Error(),
			ResponseTime: responseTime,
			RequestID:    req.RequestID,
		}
	}
	defer httpResp.Body.Close()

	// Read response body
	bodyBytes, err := io.ReadAll(httpResp.Body)
	if err != nil {
		responseTime := time.Since(startTime)
		return nil, &HTTPClientError{
			Message:      fmt.Sprintf("failed to read response body: %v", err),
			StatusCode:   httpResp.StatusCode,
			ResponseTime: responseTime,
			RequestID:    req.RequestID,
		}
	}

	responseTime := time.Since(startTime)

	// Parse response body
	var responseData interface{}
	contentType := httpResp.Header.Get("Content-Type")
	
	if strings.Contains(contentType, "application/json") {
		if len(bodyBytes) > 0 {
			if err := json.Unmarshal(bodyBytes, &responseData); err != nil {
				responseData = string(bodyBytes)
			}
		}
	} else {
		responseData = string(bodyBytes)
	}

	// Build response headers
	headers := make(map[string]string)
	for k, v := range httpResp.Header {
		if len(v) > 0 {
			headers[k] = v[0]
		}
	}

	// Create response
	response := &HTTPResponse{
		StatusCode:   httpResp.StatusCode,
		Data:         responseData,
		Headers:      headers,
		Success:      httpResp.StatusCode >= 200 && httpResp.StatusCode < 300,
		ResponseTime: responseTime,
		RequestID:    req.RequestID,
		ContentType:  contentType,
		RawBody:      bodyBytes,
	}

	// Cache successful GET responses
	if req.Method == GET && response.Success && c.config.EnableCache {
		c.cache.Set(req.URL, req.Headers, response)
	}

	// Record metrics
	if response.Success {
		c.metrics.RecordSuccess(req.Method, responseTime)
	} else {
		c.metrics.RecordHTTPError(req.Method, response.StatusCode)
	}

	// Return error for non-successful responses
	if !response.Success {
		response.Error = fmt.Sprintf("HTTP %d: %s", response.StatusCode, httpResp.Status)
		return response, &HTTPClientError{
			Message:      response.Error,
			StatusCode:   response.StatusCode,
			ResponseData: responseData,
			ResponseTime: responseTime,
			RequestID:    req.RequestID,
		}
	}

	return response, nil
}

// shouldRetry determines if a request should be retried
func (c *HTTPClient) shouldRetry(err error) bool {
	if httpErr, ok := err.(*HTTPClientError); ok {
		// Retry on 5xx errors and 429 (rate limiting)
		return httpErr.StatusCode == 429 || httpErr.StatusCode >= 500
	}
	
	// Retry on network errors
	return strings.Contains(err.Error(), "timeout") ||
		strings.Contains(err.Error(), "connection") ||
		strings.Contains(err.Error(), "network")
}

// Get performs a GET request
func (c *HTTPClient) Get(url string, options ...RequestOption) (*HTTPResponse, error) {
	req := &HTTPRequest{
		Method:    GET,
		URL:       c.buildURL(url),
		Headers:   c.buildHeaders(nil),
		RequestID: c.generateRequestID(),
		Context:   context.Background(),
	}

	// Apply options
	for _, option := range options {
		option(req)
	}

	return c.executeRequest(req)
}

// Post performs a POST request
func (c *HTTPClient) Post(url string, body interface{}, options ...RequestOption) (*HTTPResponse, error) {
	req := &HTTPRequest{
		Method:    POST,
		URL:       c.buildURL(url),
		Headers:   c.buildHeaders(nil),
		Body:      body,
		RequestID: c.generateRequestID(),
		Context:   context.Background(),
	}

	// Apply options
	for _, option := range options {
		option(req)
	}

	return c.executeRequest(req)
}

// Put performs a PUT request
func (c *HTTPClient) Put(url string, body interface{}, options ...RequestOption) (*HTTPResponse, error) {
	req := &HTTPRequest{
		Method:    PUT,
		URL:       c.buildURL(url),
		Headers:   c.buildHeaders(nil),
		Body:      body,
		RequestID: c.generateRequestID(),
		Context:   context.Background(),
	}

	// Apply options
	for _, option := range options {
		option(req)
	}

	return c.executeRequest(req)
}

// Delete performs a DELETE request
func (c *HTTPClient) Delete(url string, options ...RequestOption) (*HTTPResponse, error) {
	req := &HTTPRequest{
		Method:    DELETE,
		URL:       c.buildURL(url),
		Headers:   c.buildHeaders(nil),
		RequestID: c.generateRequestID(),
		Context:   context.Background(),
	}

	// Apply options
	for _, option := range options {
		option(req)
	}

	return c.executeRequest(req)
}

// Patch performs a PATCH request
func (c *HTTPClient) Patch(url string, body interface{}, options ...RequestOption) (*HTTPResponse, error) {
	req := &HTTPRequest{
		Method:    PATCH,
		URL:       c.buildURL(url),
		Headers:   c.buildHeaders(nil),
		Body:      body,
		RequestID: c.generateRequestID(),
		Context:   context.Background(),
	}

	// Apply options
	for _, option := range options {
		option(req)
	}

	return c.executeRequest(req)
}

// RequestOption represents a request option function
type RequestOption func(*HTTPRequest)

// WithHeaders adds headers to the request
func WithHeaders(headers map[string]string) RequestOption {
	return func(req *HTTPRequest) {
		for k, v := range headers {
			req.Headers[k] = v
		}
	}
}

// WithQueryParams adds query parameters to the request
func WithQueryParams(params map[string]string) RequestOption {
	return func(req *HTTPRequest) {
		if req.QueryParams == nil {
			req.QueryParams = make(map[string]string)
		}
		for k, v := range params {
			req.QueryParams[k] = v
		}
	}
}

// WithContext sets the context for the request
func WithContext(ctx context.Context) RequestOption {
	return func(req *HTTPRequest) {
		req.Context = ctx
	}
}

// WithTimeout sets a custom timeout for the request
func WithTimeout(timeout time.Duration) RequestOption {
	return func(req *HTTPRequest) {
		req.Timeout = timeout
	}
}

// WithRequestID sets a custom request ID
func WithRequestID(requestID string) RequestOption {
	return func(req *HTTPRequest) {
		req.RequestID = requestID
	}
}

// GetMetrics returns HTTP client metrics
func (c *HTTPClient) GetMetrics() HTTPMetricsSnapshot {
	return c.metrics.Snapshot()
}

// ClearCache clears the request cache
func (c *HTTPClient) ClearCache() {
	c.cache.Clear()
}

// RequestCache provides simple caching for GET requests
type RequestCache struct {
	items map[string]*cacheItem
	ttl   time.Duration
}

type cacheItem struct {
	response   *HTTPResponse
	expiration time.Time
}

// NewRequestCache creates a new request cache
func NewRequestCache(ttl time.Duration) *RequestCache {
	cache := &RequestCache{
		items: make(map[string]*cacheItem),
		ttl:   ttl,
	}
	
	// Start cleanup goroutine
	go cache.cleanup()
	
	return cache
}

// Get retrieves a cached response
func (c *RequestCache) Get(url string, headers map[string]string) *HTTPResponse {
	key := c.buildKey(url, headers)
	
	item, exists := c.items[key]
	if !exists || time.Now().After(item.expiration) {
		delete(c.items, key)
		return nil
	}
	
	return item.response
}

// Set stores a response in cache
func (c *RequestCache) Set(url string, headers map[string]string, response *HTTPResponse) {
	key := c.buildKey(url, headers)
	
	c.items[key] = &cacheItem{
		response:   response,
		expiration: time.Now().Add(c.ttl),
	}
}

// Clear clears all cached items
func (c *RequestCache) Clear() {
	c.items = make(map[string]*cacheItem)
}

// buildKey builds a cache key from URL and headers
func (c *RequestCache) buildKey(url string, headers map[string]string) string {
	// Simple key generation - in production, you might want something more sophisticated
	return fmt.Sprintf("%s:%v", url, headers)
}

// cleanup removes expired items from cache
func (c *RequestCache) cleanup() {
	ticker := time.NewTicker(time.Minute)
	defer ticker.Stop()
	
	for range ticker.C {
		now := time.Now()
		for key, item := range c.items {
			if now.After(item.expiration) {
				delete(c.items, key)
			}
		}
	}
}

// HTTPMetrics tracks HTTP client metrics
type HTTPMetrics struct {
	totalRequests    int64
	successfulRequests int64
	failedRequests   int64
	totalResponseTime time.Duration
	errorsByStatus   map[int]int64
	errorsByType     map[string]int64
}

// NewHTTPMetrics creates new HTTP metrics
func NewHTTPMetrics() *HTTPMetrics {
	return &HTTPMetrics{
		errorsByStatus: make(map[int]int64),
		errorsByType:   make(map[string]int64),
	}
}

// RecordSuccess records a successful request
func (m *HTTPMetrics) RecordSuccess(method HTTPMethod, responseTime time.Duration) {
	m.totalRequests++
	m.successfulRequests++
	m.totalResponseTime += responseTime
}

// RecordError records an error
func (m *HTTPMetrics) RecordError(method HTTPMethod, err error) {
	m.totalRequests++
	m.failedRequests++
	m.errorsByType[err.Error()]++
}

// RecordHTTPError records an HTTP error
func (m *HTTPMetrics) RecordHTTPError(method HTTPMethod, statusCode int) {
	m.totalRequests++
	m.failedRequests++
	m.errorsByStatus[statusCode]++
}

// Snapshot returns a snapshot of current metrics
func (m *HTTPMetrics) Snapshot() HTTPMetricsSnapshot {
	avgResponseTime := time.Duration(0)
	if m.totalRequests > 0 {
		avgResponseTime = m.totalResponseTime / time.Duration(m.totalRequests)
	}

	return HTTPMetricsSnapshot{
		TotalRequests:       m.totalRequests,
		SuccessfulRequests:  m.successfulRequests,
		FailedRequests:      m.failedRequests,
		AverageResponseTime: avgResponseTime,
		SuccessRate:         float64(m.successfulRequests) / float64(m.totalRequests) * 100,
		ErrorsByStatus:      m.errorsByStatus,
		ErrorsByType:        m.errorsByType,
	}
}

// HTTPMetricsSnapshot represents a snapshot of HTTP metrics
type HTTPMetricsSnapshot struct {
	TotalRequests       int64            `json:"total_requests"`
	SuccessfulRequests  int64            `json:"successful_requests"`
	FailedRequests      int64            `json:"failed_requests"`
	AverageResponseTime time.Duration    `json:"average_response_time"`
	SuccessRate         float64          `json:"success_rate"`
	ErrorsByStatus      map[int]int64    `json:"errors_by_status"`
	ErrorsByType        map[string]int64 `json:"errors_by_type"`
}

// API client base class
type APIClient struct {
	httpClient *HTTPClient
	baseURL    string
	apiVersion string
}

// NewAPIClient creates a new API client
func NewAPIClient(baseURL, apiVersion string, config HTTPClientConfig) *APIClient {
	config.BaseURL = baseURL
	return &APIClient{
		httpClient: NewHTTPClient(config),
		baseURL:    baseURL,
		apiVersion: apiVersion,
	}
}

// buildEndpoint builds an API endpoint URL
func (c *APIClient) buildEndpoint(endpoint string) string {
	return fmt.Sprintf("/api/%s/%s", c.apiVersion, endpoint)
}

// Get performs a GET request to the API
func (c *APIClient) Get(endpoint string, options ...RequestOption) (*HTTPResponse, error) {
	return c.httpClient.Get(c.buildEndpoint(endpoint), options...)
}

// Post performs a POST request to the API
func (c *APIClient) Post(endpoint string, body interface{}, options ...RequestOption) (*HTTPResponse, error) {
	return c.httpClient.Post(c.buildEndpoint(endpoint), body, options...)
}

// Put performs a PUT request to the API
func (c *APIClient) Put(endpoint string, body interface{}, options ...RequestOption) (*HTTPResponse, error) {
	return c.httpClient.Put(c.buildEndpoint(endpoint), body, options...)
}

// Delete performs a DELETE request to the API
func (c *APIClient) Delete(endpoint string, options ...RequestOption) (*HTTPResponse, error) {
	return c.httpClient.Delete(c.buildEndpoint(endpoint), options...)
}

// Example usage demonstrates how to use the HTTP client
func ExampleUsage() {
	// Create HTTP client
	config := DefaultHTTPClientConfig()
	config.BaseURL = "https://api.example.com"
	config.EnableLogging = true
	
	client := NewHTTPClient(config)

	// Make GET request
	resp, err := client.Get("/users/1")
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}
	
	fmt.Printf("GET Response: %+v\n", resp.Data)

	// Make POST request
	userData := map[string]interface{}{
		"name":  "John Doe",
		"email": "john@example.com",
	}
	
	resp, err = client.Post("/users", userData)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}
	
	fmt.Printf("POST Response: %+v\n", resp.Data)

	// Use API client
	apiClient := NewAPIClient("https://api.example.com", "v1", config)
	resp, err = apiClient.Get("users/1")
	if err != nil {
		fmt.Printf("API Error: %v\n", err)
		return
	}
	
	fmt.Printf("API Response: %+v\n", resp.Data)

	// Get metrics
	metrics := client.GetMetrics()
	fmt.Printf("HTTP Metrics: %+v\n", metrics)
}
